[GeoMechanicsApplication] Add a time incrementor that models a collection of prescribed time increments

avdg81 opened the pull request:
**📝 Description**
Added a time incrementor that implements the time incrementor interface for the case where we have a collection of prescribed time increments. This class is probably one of the simplest (and most fundamental) ways to step through time. We will use it when developing the time loop executor. (But we may consider to also make it accessible through the `ProjectParameters.json` file.)

avdg81 committed:
Minimal class for prescribed time increments

avdg81 committed:
Ask for a next step when the increment list is not empty

avdg81 committed:
Added an iterator that points in the list of increments

Whenever a time step has finished and its result has been processed, we
advance the iterator. In this way, when we have a list of N time
increments and we process N time step results, the next query whether a
next step must be carried out will yield false.

Also removed two empty statements.

avdg81 committed:
When asked whether we want to retry a step, always return true

avdg81 committed:
Only retry a step when it's the first attempt (i.e. cycle)

avdg81 committed:
Return the time increment pointed to by the internal iterator

avdg81 committed:
Throw an exception when any of the time increments is negative

avdg81 committed:
Throw an exception when any of the time increments is negative

avdg81 committed:
Throw an exception when attempting to get increment beyond end

avdg81 committed:
When end of increment list is achieved never do another step

avdg81 committed:
Without post time step execution we always get same increment

avdg81 committed:
All unit tests so far assume a converged end state

avdg81 committed:
Don't want a next step when previous one did not converge

rfaasse commented on applications/GeoMechanicsApplication/tests/cpp_tests/test_prescribed_time_incrementor.cpp:
```diff
@@ -0,0 +1,143 @@
+
+KRATOS_TEST_CASE_IN_SUITE(NoNextTimeStepWhenPrescribedTimeIncrementorHasEmptyList, KratosGeoMechanicsFastSuite)
+{
+    std::vector<double> increments;
```
In all tests, the increments vector can be const std::vector<double> :)

avdg81 replied:
Good catch. In fact, there were a few more variables that could also be made `const`. I have changed those as well.

rfaasse commented on applications/GeoMechanicsApplication/custom_workflows/prescribed_time_incrementor.cpp:
```diff
@@ -0,0 +1,58 @@
+    return *mPos;
+}
+
+void PrescribedTimeIncrementor::PostTimeStepExecution(const Kratos::TimeStepEndState& rResultantState)
```
Just for my understanding: This is split from the GetIncrement() function in the interface, such that the adaptive time incrementor can re-try a time-step right?

avdg81 replied:
Yes, this API is there to give the incrementor the opportunity to adjust the next increment if needed. `GetIncrement()` is supposed to do nothing more than fetching the increment to be applied.

rfaasse reviewed (dismissed):
Looks very clear and complete! Nice to see TDD happening more and more 😄 

Just flagged a a couple of missing 'consts' but nothing major.

avdg81 committed:
Attempt to fix the broken build

- Ensure that a returned value is assigned to a variable when the
  function is marked `[[nodiscard]]`.
- Removed two empty statements.

avdg81 committed:
Fixed another build error due to an unused variable

The unused variable was added to suppress a warning due to the
`[[nodiscard]]` attribute of a member function. To avoid another warning
due to the unused variable, the dummy variable has now been decorated
with the `[[maybe_unused]]` attribute.

avdg81 committed:
Made a few more variables in unit tests `const`

rfaasse reviewed (approved):
Looks good to me!
